lib/consumerGroup.js

```diff
@@ -762,6 +733,144 @@ ConsumerGroup.prototype.removeTopics = function (topics, cb) {
+    return p;
+  });
+};
+
+ConsumerGroup.prototype.buildPayloadMap = function (payloads) {
+  const payloadMap = {};
+  payloads.forEach(({ topic, partition, offset }) => {
+    payloadMap[topic] = payloadMap[topic] || {};
+    payloadMap[topic][partition] = offset;
+  });
+  return payloadMap;
+};
+
+/*
+ * Update offset info in current payloads
+ * @param {Object} Topic-partition-offset
+ * @param {Boolean} Don't commit when initing consumer
+ */
+ConsumerGroup.prototype.updateOffsets = function (topics, initing) {
+  this.topicPayloads.forEach(p => {
+    if (!_.isEmpty(topics[p.topic]) && topics[p.topic][p.partition] !== undefined) {
+      var offset = topics[p.topic][p.partition];
+      if (offset === -1) offset = 0;
+      if (!initing) p.offset = offset + 1;
+      else p.offset = offset;
+      // Update the map
+      this.needToCommit = true;
+    }
+    this.payloadMap[p.topic] = this.payloadMap[p.topic] || {};
+    this.payloadMap[p.topic][p.partition] = p.offset;
+  });
+
+  if (this.options.autoCommit && !initing) {
+    this.autoCommit(false, function (err) {
+      err && logger.debug('auto commit offset', err);
+    });
+  }
+};
+
+ConsumerGroup.prototype._onFetchDone = function (topics) {
+  this.updateOffsets(topics);
+  if (--this._pendingFetches > 0) {
+    return;
+  }
+  this._isFetchPending = false;
+  if (!this.paused) {
+    setImmediate(() => this.fetch());
+  }
+};
+
+ConsumerGroup.prototype._resetFetchState = function () {
+  this._pendingFetches = 0;
+  this._isFetchPending = false;
+};
+
+ConsumerGroup.prototype._onFetchProcessing = function () {
+  this._pendingFetches++;
+};
+
+ConsumerGroup.prototype.fetch = function () {
+  if (!this.ready || this.rebalancing || this.paused || this.closing) {
+    return;
+  }
+  if (this._isFetchPending) {
+    return;
+  }
+  this._isFetchPending = true;
+
+  this.client.sendFetchRequest(
+    this,
+    this.topicPayloads,
+    this.options.fetchMaxWaitMs,
+    this.options.fetchMinBytes,
+    this.options.maxTickMessages
+  );
+};
+
+ConsumerGroup.prototype.setOffset = function (topic, partition, offset) {
+  this.topicPayloads.every(function (p) {
+    // eslint-disable-next-line eqeqeq
+    if (p.topic === topic && p.partition == partition) {
+      p.offset = offset;
+      return false;
+    }
+    return true;
+  });
+};
+
+ConsumerGroup.prototype.pause = function () {
+  this.paused = true;
+};
+
+ConsumerGroup.prototype.resume = function () {
+  this.paused = false;
+  this.fetch();
+};
+
+function autoCommit (force, cb) {
+  if (arguments.length === 1) {
+    cb = force;
+    force = false;
+  }
+
+  if (!force) {
+    if (this.committing) return cb(null, 'Offset committing');
+    if (!this.needToCommit) return cb(null, 'Commit not needed');
+  }
+
+  this.needToCommit = false;
+  this.committing = true;
+  setTimeout(
+    function () {
+      this.committing = false;
+    }.bind(this),
+    this.options.autoCommitIntervalMs
+  ).unref();
```

Comment:
Hey @hyperlink `unref` causes an error in an Electron environment.
```
Uncaught TypeError: setTimeout(...).unref is not a function
    at ConsumerGroup.autoCommit
``` 

Is there an alternative way to clean up the timeout reference?

Reply:
I suppose we can check to see if `unref` is available and then call it if it's there.